packages/netsuite-adapter/src/change_validators/unique_fields.ts

```diff
@@ -0,0 +1,235 @@
+      3. Update the getters map.
+      4. Update getRestrictedType to match the new type as well. */
+
+type RestrictedTypes = typeof SAVED_SEARCH_TYPE | typeof FINANCIAL_LAYOUT_TYPE | typeof CUSTOM_RECORD_FIELD_TYPE
+
+const savedSearchGetters: RestrictedTypeGetters = {
+  getChangeRestrictedField: change => getChangeNestedField(change, FIELD_DEFAULT_NAME),
+  getSourceRestrictedFields: params => getNestedField(params, FIELD_DEFAULT_NAME),
+  getMessage: () => 'A Saved Search with that title already exists',
+  getDetailedMessage: field => `Can't deploy the Saved Search: ${field} because there is already a Saved Search with that title in the target environment.`
+    + ' Please select a Saved Search with a unique title.',
+}
+
+const financialLayoutGetters: RestrictedTypeGetters = {
+  getChangeRestrictedField: change => getChangeNestedField(change, NAME_FIELD),
+  getSourceRestrictedFields: params => getNestedField(params, NAME_FIELD),
+  getMessage: () => 'A Financial Layout with that name already exists',
+  getDetailedMessage: (field: string) => `Can't deploy the Financial Layout: ${field} because there is already a Financial Layout with that name in the target environment.`
+    + ' Please select a Financial Layout with a unique name.',
+}
+
+const customRecordGetters: RestrictedTypeGetters = {
+  getChangeRestrictedField: change => getChangeNestedField(change, SCRIPT_ID),
+  getSourceRestrictedFields: getCustomRecordRestrictedData,
+  getMessage: () => 'A Custom Record Type Field with that ID already exists',
+  getDetailedMessage: field => `Can't deploy the Custom Record Type Field: ${field} because there is already a Custom Record Type Field with that ID in the target environment.`
+    + ' Please select a Custom Record Type Field with a unique ID.',
+}
+
+const restrictedTypesGettersMap: Record<RestrictedTypes, RestrictedTypeGetters> = {
+  savedSearch: savedSearchGetters,
+  financialLayout: financialLayoutGetters,
+  customRecordField: customRecordGetters,
+}
+
+const getRestrictedType = (elemID: ElemID, isChange = true): RestrictedTypes | undefined => {
+  if (elemID.idType === INSTANCE_TYPE) {
+    if (elemID.typeName === SAVED_SEARCH) {
+      return SAVED_SEARCH_TYPE
+    }
+
+    if (elemID.typeName === FINANCIAL_LAYOUT) {
+      return FINANCIAL_LAYOUT_TYPE
+    }
+  } else if ((isChange && elemID.idType === FIELD_TYPE) || (!isChange && elemID.idType === TYPE_TYPE)) {
+    if (elemID.typeName.startsWith(CUSTOM_RECORD_ID_PREFIX)) {
+      return CUSTOM_RECORD_FIELD_TYPE
+    }
+  }
+
+  return undefined
+}
+
+/* From this place forward the functions are general.
+   Change them carefully only if the new type's logic requires that. */
+
+const getTypeToDataMap = <T extends ElemID | ChangeDataType>(
+  elements: T[],
+  isChange = true,
+  typeToChanges?: Map<RestrictedTypes, ChangeDataType[]>
+):
+  Map<RestrictedTypes, T[]> => {
+  const restrictedGroups: Map<RestrictedTypes, T[]> = new Map()
```

Comment:
why not using a `Record`? you can initiate it with empty lists so the condition in line 150 will be unnecessary: 
```suggestion
  const restrictedGroups: Record<RestrictedTypes, T[]> = {
       savedSearch: [],
       financialLayout: [],
       customRecordField: [],
  }
```

Reply:
In general, anywhere that uses `Map<RestrictedTypes, ...>` should be replaced with `Record<...>` IMO - because the keys are well-defined.

Reply:
The places I used Map instead of Record are ones where we don't necessarily need a mapping of **all** the RestrictedTypes.
In addition, using Record requires initialization with all the RestrictedTypes, which extends the areas in the code that require changes when adding a new restricted type. I wanted to avoid that and make the code as 'generic' as I could.

WDYT?

Reply:
IMO It's not bad that adding a new restrictedType requires changes in the code - it's an advantage that TS force us to add it, and this way prevent missing some of the places. You can still avoid code duplication by having a function that generates an empty Record.